### PR TITLE
Add nightly dex2oat compiler

### PR DESCRIPTION
**Pull request by kevinjeon-g**
This change adds dex2oat for Android as a nightly compiler. The zip file containing the compiler and other necessary artifacts is pulled from the latest AOSP build using the fetch_art_release_from_head.sh script.

The script has been uploaded into the android-java/ directory, and script.py has been modified to make the 'fetch' config optional.

### CODE REVIEW

File: bin/yaml/android-java/fetch_art_release_from_head.sh [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# Fetches ART binaries and ART bootclasspath jars from AOSP head.
+# Saves the result at `art_release.zip`.
+
+set -e
```
Comment (mattgodbolt) [inline bin/yaml/android-java/fetch_art_release_from_head.sh:6]:
```suggestion
set -euo pipefail
```

(if this is not just a copy paste...)

File: bin/yaml/android-java/fetch_art_release_from_head.sh [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+
+URL=https://ci.android.com/builds/latest/branches/aosp-main/targets/aosp_arm64-trunk_staging-userdebug/view/BUILD_INFO
+RURL=$(curl -Ls -o /dev/null -w "%{url_effective}" "${URL}")
+wget "${RURL%/view/BUILD_INFO}/raw/art_release.zip"
```
Comment (mattgodbolt) [inline bin/yaml/android-java/fetch_art_release_from_head.sh:10]: Using both `curl` _and_ `wget` seems excessive. We definitely have `curl` around but not necessarily `wget`

File: bin/yaml/android-java/fetch_art_release_from_head.sh [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+# Fetches ART binaries and ART bootclasspath jars from AOSP head.
+# Saves the result at `art_release.zip`.
+
+set -e
+
+URL=https://ci.android.com/builds/latest/branches/aosp-main/targets/aosp_arm64-trunk_staging-userdebug/view/BUILD_INFO
```
Comment (mattgodbolt) [inline bin/yaml/android-java/fetch_art_release_from_head.sh:8]: will this binary run on an x86 linux machine?

File: bin/yaml/android-java/fetch_art_release_from_head.sh (hunk 1 above)
Comment (kevinjeon-g) [reply bin/yaml/android-java/fetch_art_release_from_head.sh]: This is a copy-paste, but the added flags work. Done!

File: bin/yaml/android-java/fetch_art_release_from_head.sh (hunk 2 above)
Comment (kevinjeon-g) [reply bin/yaml/android-java/fetch_art_release_from_head.sh]: That makes sense—the script has been updated to use curl --location instead.

File: bin/yaml/android-java/fetch_art_release_from_head.sh (hunk 3 above)
Comment (kevinjeon-g) [reply bin/yaml/android-java/fetch_art_release_from_head.sh:8]: Yes, the binary is built for x86_64. The arm64 here refers to the compiler output, not that the binary runs on an arm64 machine.

File: bin/yaml/android-java.yaml [hunk 4]
```diff
@@ -10,3 +10,20 @@ compilers:
     url: https://dl.google.com/android/maven2/com/android/tools/r8/{name}/r8-{name}.jar
     targets:
       - 8.1.56
+
+  nightly:
+    if: nightly
+    type: script
+    dir: "{name}"
+    fetch:
```
Comment (mattgodbolt) [inline bin/yaml/android-java.yaml:18]: is this fetch meant to be empty?

File: bin/lib/installable/script.py [hunk 5]
```diff
@@ -14,7 +14,7 @@ def __init__(self, install_context: InstallationContext, config: Dict[str, Any])
         super().__init__(install_context, config)
         self.install_path = self.config_get("dir")
         self.install_path_symlink = self.config_get("symlink", False)
-        self.fetch = self.config_get("fetch")
+        self.fetch = self.config_get("fetch", [])
```
Comment (mattgodbolt) [inline bin/lib/installable/script.py:17]: Was this change required?

**Review by mattgodbolt — APPROVED** [review]
As I've taken so long,  I'm going to merge anyway! sorry for the delay

**Comment by mattgodbolt** [comment]
Installed ok :) 
```
2023-12-23 22:26:38,107 lib.ce_install  INFO     compilers/nightly/dex2oat dex2oat-latest installed OK
1 packages installed OK, 0 skipped, and 0 failed installation
```

File: bin/lib/installable/script.py (hunk 5 above)
Comment (kevinjeon-g) [reply bin/lib/installable/script.py:17]: When I was testing, listing nightly compilers (and trying to install) would fail because there wasn't a fetch: for the dex2oat config. I had added the default empty list arg to avoid faking a fetch arg!

File: bin/yaml/android-java.yaml (hunk 4 above)
Comment (kevinjeon-g) [reply bin/yaml/android-java.yaml:18]: Ahh, this line was meant to be deleted... I had added it while testing out the change to script.py, but had forgotten to clean it up again. Sorry about that! I see that you fixed it in a later change—thank you!
